### ISSUE TITLE
No warning with `set -e` and eval inside string interpolation

### ISSUE THREAD

**Issue by toothbrush** [koalaman/shellcheck#2478]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

I did search existing issues, but wasn't sure of the correct terminology to use – so if this is a dupe, i apologise.

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/usr/bin/env bash
set -e
echo "This is a string with some $(false) content"
echo "I don't want to get here"
```

#### Here's what shellcheck currently says:

No warnings at all!

#### Here's what I wanted or expected to see:

It would be great if `shellcheck` could warn the user that a nonzero exit code inside a string interpolation with evaluation (or whatever one calls `"Some string with $(command output)."`) is a bit of a footgun – the `command` may fail, but the script won't stop executing as you'd perhaps expect with `set -e`.  The simple workaround that can be suggested, which i use in my scripts, is as follows: extract the command execution into a temporary variable set before the place where it's used.

```sh
#!/usr/bin/env bash
set -e
some_string=$(false) # exits early because of set -e
echo "This is a string with some ${some_string} content" # never evaluated, script aborts
```

#### Real-world example

I recently got bitten by this and would've liked a `shellcheck` warning. Basically, i am attempting to configure a correct build path, but this stanza depends on the user having the `mysql@5.7` Homebrew package installed.  The problem is that when a user **does not** have the right package installed, `brew --installed ...` exits nonzero, but `bundle config` is still called, just with an empty string.  It is preferable to fail rather than set invalid config silently, so i consider this a footgun.

```
#!/usr/bin/env bash
set -e
bundle config \
  --local build.mysql2 \
  --with-mysql-config="$(brew --prefix --installed mysql@5.7)/bin/mysql_config" 
```

I hope my description makes sense, and isn't a duplicate!  Thanks for making `shellcheck`! 🙌

**Comment by koalaman** [comment]
Have you tried the optional `check-extra-masked-returns` in 0.8.0?

With e.g. 

```
#!/usr/bin/env bash
# shellcheck enable=check-extra-masked-returns
set -e
echo "This is a string with some $(false) content"
echo "I don't want to get here"
```
you get

```
In foo line 4:
echo "This is a string with some $(false) content"
                                   ^---^ SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).
```

**Comment by toothbrush** [comment]
Oh, amazing!  That's exactly what i was looking for!  Sorry – i'm not used to changing the default-enabled shellchecks, because we also run it in CI a lot and i try not to stray too much from default.

I'll consider adding that to my local config, thank you!

Is there any scope for promoting that to the "default on" set of shellchecks?

Thanks for your quick response 🙏

**Comment by bjwrk** [comment]
Wow, this is great! What's the rationale for "off-by-default" for this one @koalaman? This is a pretty big footgun in bash.

**Comment by toothbrush** [comment]
Just in case others come here looking, i dropped the following into `~/.shellcheckrc`:

```
enable=check-extra-masked-returns
```

**Comment by koalaman** [comment]
@toothbrush You can also add it to a `.shellcheckrc` in your project root dir if you want all users (including CI) to use it for that project.

@bjwrk The main reason is that it has a very high trigger rate for code that realistically speaking is fine. Some people love this kind of 100% completion, while others think it's pedantic and unhelpful since they're already checking the exit status of all commands they care about.
